analyzers/src/SonarAnalyzer.Common/Rules/InsteadOfAnyBase.cs

```diff
@@ -0,0 +1,124 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class InsteadOfAnyBase<TSyntaxKind, TInvocationExpression> : SonarDiagnosticAnalyzer
+    where TSyntaxKind : struct
+    where TInvocationExpression : SyntaxNode
+{
+    private const string ExistsDiagnosticId = "S6605"; // Collection-specific "Exists" method should be used instead of the "Any" extension.
+    private const string ContainsDiagnosticId = "S6617"; // Collection-specific "Contains" method should be used instead of the "Any" extension.
+    private const string MessageFormat = "Collection-specific \"{0}\" method should be used instead of the \"Any\" extension.";
+
+    private readonly DiagnosticDescriptor existsRule;
+    private readonly DiagnosticDescriptor containsRule;
+    private ImmutableArray<KnownType> ExistsTypes { get; } = ImmutableArray.Create(
+        KnownType.System_Array,
+        KnownType.System_Collections_Immutable_ImmutableList_T);
+
+    private ImmutableArray<KnownType> ContainsTypes { get; } = ImmutableArray.Create(
+        KnownType.System_Collections_Generic_HashSet_T,
+        KnownType.System_Collections_Generic_SortedSet_T);
+
+    protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(existsRule, containsRule);
+
+    protected abstract bool IsSimpleEqualityCheck(TInvocationExpression node, SemanticModel model);
+    protected abstract SyntaxNode GetArgumentExpression(TInvocationExpression invocation, int index);
+    protected abstract bool AreValidOperands(string lambdaVariable, SyntaxNode first, SyntaxNode second);
+
+    protected InsteadOfAnyBase()
+    {
+        existsRule = Language.CreateDescriptor(ExistsDiagnosticId, MessageFormat);
+        containsRule = Language.CreateDescriptor(ContainsDiagnosticId, MessageFormat);
+    }
```

Comment:
If the previous comment is done, the constructor can be safely deleted.

Reply:
See comment above